scipy/stats/stats.py

```diff
@@ -7189,9 +7198,316 @@ def mannwhitneyu(x, y, use_continuity=True, alternative=None):
+        '''Probability mass function'''
+        return self.f(m, n, k) / special.binom(m + n, m)
+
+    def cdf(self, k, m, n):
+        '''Cumulative distribution function'''
+        return (np.sum([self.f(m, n, i) for i in range(0, k+1)])
+                / special.binom(m + n, m))
+
+    def sf(self, k, m, n):
+        '''Survival function'''
+        n_choose = special.binom(m + n, m)
+        # I think this should be k+1, but k makes it match expected result : /
+        F = np.sum([self.f(m, n, i) for i in range(0, k)])
+        # consequently, cdf + sf > 1; we're double counting pmf(k, m, n)
+
+        # the thought was that we could do (n_choose - F) with integer
+        # arithmetic, but it probably doesn't matter because n_choose can
+        # easily exceed the maximum int64
+        return (n_choose - F) / n_choose
+
+    def f(self, m, n, k):
+        '''Convenience function of [3] Theorem 2.5'''
+        # If necessary, expand the array that remembers PMF values
+        # could probably use `np.pad` but I'm not sure it would save code
+        shape_old = np.array(self.fmnks.shape)
+        shape_new = np.array((m+1, n+1, k+1))
+        if np.any(shape_new > shape_old):
+            shape = np.maximum(shape_old, shape_new)
+            fmnks = -np.ones(shape)            # create the new array
+            m0, n0, k0 = shape_old
+            fmnks[:m0, :n0, :k0] = self.fmnks  # copy rememered values
+            self.fmnks = fmnks
+
+        # Call the function that does all the real work
+        return self._f(m, n, k)
+
+    def _f(self, m, n, k):
+        '''Recursive implementation function of [3] Theorem 2.5'''
+        # TODO: vectorize w.r.t. k
+
+        fmnks = self.fmnks  # for convenience
+
+        # [3] Theorem 2.5 Line 1
+        if k < 0 or m < 0 or n < 0 or k > m*n:
+            return 0
+
+        # if already calculated, return the value
+        if fmnks[m, n, k] >= 0:
+            return fmnks[m, n, k]
+
+        if k == 0 and m >= 0 and n >= 0:  # [3] Theorem 2.5 Line 2
+            fmnk = 1
+        else:   # [3] Theorem 2.5 Line 3 / Equation 3
+            fmnk = self.f(m-1, n, k-n)  +  self.f(m, n-1, k)
+
+        fmnks[m, n, k] = fmnk  # remember result
+
+        return fmnk
+
+
+# Maintain state for faster repeat calls to mannwhitneyu2 w/ method='exact'
+_mwu_state = _MWU()  #
+
+
+def _get_mwu_z(U, n1, n2, ranks, axis=0, continuity=True):
+    '''Standardized MWU statistic'''
+    # Follows mannwhitneyu2 [2]
+    n = n1 + n2
+    c = -0.5 * continuity
+    m_u = n1 * n2 / 2
+
+    # Tie correction. scipy.stats.tiecorrect is not vectorized; this is.
+    _, t = np.unique(ranks, return_counts=True, axis=-1)
+    s = np.sqrt(n1*n2/12 * (n1 + n2 + 1 - (t**3 - t).sum(axis=-1)/(n*(n-1))))
+
+    z = (U + c - m_u) / s
+    return z
+
+
+mwu_result = make_dataclass("MannWhitneyUResult", ("statistic", "pvalue"))
+
+
+def mannwhitneyu2(x, y, continuity=True, alternative=None, axis=0,
```

Comment:
@ev-br I wonder if this approach here addresses some of your remaining comments about the `mann_whitney_u` above (which I've left in for comparison).  Some advantages are that the calculation of the test-statistic and asymptotic p-value is vectorized for N-d input (addresses gh-12837), and the calculation of the exact test statistic is broken into a separate class that memoizes the results for reuse. This also makes it much faster as the input size grows. To make review easier, it's written to align obviously with references (Wikipedia for the statistic calculation, [this reference](https://pure.tue.nl/ws/files/1610937/467722.pdf) for exact distribution of U under the null hypothesis).

Some of the choices (e.g. default `alternative=None`) are temporary to pass the tests from the original `mannwhitneyu`. It's still in draft form, but just thought I'd put this up here to see if we want to head in this direction.